iroh-ctl/src/status.rs

```diff
@@ -0,0 +1,114 @@
+use std::collections::HashMap;
+use std::io::{stdout, Write};
+
+use crossterm::terminal::{Clear, ClearType};
+use crossterm::{
+    cursor,
+    style::{self, Attribute, Color, Stylize},
+    QueueableCommand,
+};
+use futures::StreamExt;
+
+use anyhow::Result;
+use iroh_rpc_client::{Client, RpcClientConfig, ServiceStatus};
+
+pub(crate) async fn status(watch: bool) -> Result<()> {
+    let client = Client::new(&RpcClientConfig::default()).await.unwrap();
+
+    let mut stdout = stdout();
+    if watch {
+        let status_stream = client.watch().await;
+        futures::pin_mut!(status_stream);
+        stdout.queue(Clear(ClearType::All))?;
+        while let Some(s) = status_stream.next().await {
+            stdout
+                .queue(cursor::RestorePosition)?
+                .queue(Clear(ClearType::FromCursorUp))?
+                .queue(cursor::MoveTo(0, 1))?;
+            queue_status_table(&stdout, s)?;
+            stdout
+                .queue(cursor::SavePosition)?
+                .queue(style::Print("\n"))?
+                .flush()?;
+        }
+        Ok(())
+    } else {
+        let s = client.check().await;
+        queue_status_table(&stdout, s)?;
+        stdout.flush()?;
+        Ok(())
+    }
+}
+
+fn queue_status_table(
```

Comment:
given that you recreate this in a loop, maybe an updating api would make more sense?

```rust
struct Table {
  gateway: StatusTable,
  p2p: StatusTable,
  store: StatusTable
}

impl Table {
  fn update(&mut self, statuses: HashMap<String, ServiceStatus>) {
    // maybe make the statuses have compile time keys as well, there are no dynamic services asfaik
   if let Some(status) = statuses.get("gateway") {
     self.gateway.status = status;
    }
  }

  fn print(&self, mut stdout: &std::io::Stdout) -> Result<()> {
    self.gateway.print(stdout)?;
    self.p2p.print(stdout)?;
    self.store.print(stdout)?;
  }
}

impl StatusTable {
  fn print(&self, mut stdout: &std::io::Stdout) -> Result<()> {
    // logic for printing a single table line
  }
}

```

Reply:
and then use it like this

```rust
    let mut stdout = stdout();
    let mut table = Table::default();

    if watch {
        let status_stream = client.watch().await;
        futures::pin_mut!(status_stream);
        stdout.queue(Clear(ClearType::All))?;
        while let Some(s) = status_stream.next().await {
            table.update(s);
            stdout
                .queue(cursor::RestorePosition)?
                .queue(Clear(ClearType::FromCursorUp))?
                .queue(cursor::MoveTo(0, 1))?;
            table.print(&stdout)?;
            stdout
                .queue(cursor::SavePosition)?
                .queue(style::Print("\n"))?
                .flush()?;
        }
        Ok(())
    } else {
        let s = client.check().await;
        table.update(s);
        table.print(&stdout);
        stdout.flush()?;
        Ok(())
    }
```

Reply:
This is great, thank you!